### PR TITLE
Add `transform_func` and `ylim_func` parameters to `interact()`

### PR DESCRIPTION
**Pull request by gully**
Addresses Issue #578 

Outstanding issues:
- [ ] Transformations that change the number of cadences in a lightcurve are currently not supported and would result in either a silent fail or an uncaught bokeh error.
- [x] Add a test that passes in this kwarg
- [ ] Transformations currently must preserve the `cadenceno` attribute.

### CODE REVIEW

File: lightkurve/interact.py [hunk 1]
```diff
@@ -517,6 +534,8 @@ def save_lightcurve():
                 selected_indices = np.array(tpf_source.selected.indices)
                 selected_mask = np.isin(pixel_index_array, selected_indices)
                 lc_new = tpf.to_lightcurve(aperture_mask=selected_mask)
+                if transform_func is not None:
+                    lc_new = transform_func(lc_new)
```
Comment (barentsen) [inline lightkurve/interact.py]: It looks like lines 490-494, 534-538, and perhaps 438-440 do the same thing.  Would it be good to separate this into a single helper function (`_create_lc` or whatever) so we reduce the likelihood of bugs in future updates?

**Comment by gully** [comment]
Now also addresses #577

**Comment by gully** [comment]
I think this PR is ready for review/merge, yeah?

File: lightkurve/tests/test_interact.py [hunk 2]
```diff
@@ -88,6 +88,22 @@ def test_custom_exported_filename():
         tpf[mask].interact()
 
 
+@pytest.mark.skipif(bad_optional_imports, reason="requires bokeh")
+def test_transform_and_ylim_funcs():
+    """Test the transform_func and ylim_func"""
+    import bokeh
```
Comment (barentsen) [inline lightkurve/tests/test_interact.py]: I think this line can be removed:

```suggestion

```

File: lightkurve/interact.py [hunk 3]
```diff
@@ -397,9 +403,30 @@ def show_interact_widget(tpf, notebook_url='localhost:8888',
...
+        A function that transforms the lightcurve.  The function takes in a
+        LightCurve object as input and returns a LightCurve object as output.
+        For example, the function could be to normalize the lightcurve:
+
+            transform_func = lambda lc: lc.normalize()
+
+        Or it could be more complex, such as detrending the lightcurve.  In this
+        way, the interactive selection of aperture mask can be evaluated after
+        inspection of the transformed lightcurve.  See the tutorial for some
+        limitations. The transform_func is applied before saving a fits file.
```
Comment (barentsen) [inline lightkurve/interact.py]: I'd remove the sentence `See the tutorial for some limitations.` for now, unless you can link to a specific tutorial?

**Comment by barentsen** [comment]
@gully This is very minor, but you may be able to improve the formatting of the example code in the parameters docstring slightly:
Screenshot from 2019-10-08 09-44-34

**Comment by barentsen** [comment]
@gully Hmmm, `tpf.interact()` does not appear to work for me at all right now (both on this branch and on master, using both Chrome and Firefox).  Chrome reports the following error:
```
Couldn't process kernel message TypeError: Cannot read property 'trim' of undefined
    at output (eval at append_javascript (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:35081), <anonymous>:47:39)
    at Kernel._handle_output_message (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:38184)
    at i (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:2)
    at Kernel._handle_iopub_message (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:38211)
    at Kernel._finish_ws_message (main.min.js?v=1bc4ae6ca8730320ff059b16de698691:38003)
    at main.min.js?v=1bc4ae6ca8730320ff059b16de698691:37994
```

**Comment by barentsen** [comment]
@gully `interact()` is working fine for me again, both on master and this branch, so you can ignore my previous post.

Can you have a quick look at the other comments I left above?  If you fix them we should be good to get this merged!

Finally, to enable an easy merge, can you rebase on master and edit `CHANGES.rst` to mention this new feature in the change log?  Thanks++.

**Comment by barentsen** [comment]
Ping @gully -- do you want to look into the changes requested above so we can put this puppy to bed?  (Changes needed should be fairly small.)

**Comment by gully** [comment]
OK, I think this PR is ready for merge once the tests pass.  The only slightly annoying thing is that you must preserve the number of cadences.  So for example, your transform function cannot include `lc.remove_outliers()`. Changing this requirement would be messy, since the cadence number mapping is hardcoded in several places.

**Comment by gully** [comment]
Woohoo all tests have passed and the branch is rebased on master.  ping @barentsen Feel free to review/merge.

File: CHANGES.rst [hunk 4]
```diff
@@ -8,7 +8,8 @@
 - Changed the ``SFFCorrector`` class to use the new ``RegressionCorrector``
   framework. [#613]
 
-
+- Added user customization for lightcurve processing and plotting in the
+  `.interact()` method for TargetPixelFile objects. [#600]
```
Comment (barentsen) [inline CHANGES.rst]: Let's make this more descriptive:
```Added  ``transform_func`` and ``ylim_func`` keywords to ``TargetPixelFile.interact()`` to support custom light curve transformations and y-axis limits. [#600]```

File: lightkurve/interact.py [hunk 5]
```diff
@@ -401,9 +406,28 @@ def show_interact_widget(tpf, notebook_url='localhost:8888',
...
         properly. If no protocol is supplied in the URL, e.g. if it is
         of the form "localhost:8888", then "http" will be used.
-    max_cadences : int
+    max_cadences: int
         Raise a RuntimeError if the number of cadences shown is larger than
         this value. This limit helps keep browsers from becoming unresponsive.
+    transform_func: function
+        A function that transforms the lightcurve.  The function takes in a
+        LightCurve object as input and returns a LightCurve object as output.
+        For example, the function could be to normalize the lightcurve::
```
Comment (barentsen) [inline lightkurve/interact.py]:
```suggestion
        For example, the function could be used to normalize the lightcurve::
```

**Comment by barentsen** [comment]
@gully The example code in the keyword docstrings still doesn't seem to render correctly (see six comments up).  Can you fix?

**Comment by gully** [comment]
Ahh, the docstring stems from the TargetPixelFile method, not the standalone interact.py sub-module.  Fixed the docs now, and added a nice looking Example section.

**Comment by gully** [comment]
Weird.  The macOS test is failing because the design matrix class is attempting a `.plot()`, which appears to be looking for an OS X-specific backend gui.  We saw similar problems in the past, if I recall correctly.

**Pull request merged**